PolyData

OndrejKincl opened the pull request:
As requested, I send a very simple modification which allows to handle polydata and a test which shows that it works.

Disclaimers:
- I did not know what to set as n_cells. Polydata does not really have cells. I set it to zero.
- Be warned that I am a mathematician. Not a programmer.
- Polydata also allow lines and triangle patches which I did not test.

Closes #10

ranocha commented on test/runtests.jl:
```diff
@@ -257,6 +259,54 @@ mkpath(TEST_EXAMPLES_DIR)
+    
+    # define points
+    n = 20
+    points = zeros(3,n)
+    points[1,:] .= [cos(4*pi*i/n) for i in 0:n-1]
+    points[2,:] .= [sin(4*pi*i/n) for i in 0:n-1]
+    points[3,:] .= [0.2i for i in 0:n-1]
```
Some minor formatting stuff that can help quite a bit when working in larger projects (also below). 
```suggestion
    points = zeros(3, n)
    points[1, :] .= [cos(4 * pi * i / n) for i in 0:(n - 1)]
    points[2, :] .= [sin(4 * pi * i / n) for i in 0:(n - 1)]
    points[3, :] .= [0.2 * i for i in 0:n-1]
```

ranocha commented on test/runtests.jl:
```diff
@@ -257,6 +259,54 @@ mkpath(TEST_EXAMPLES_DIR)
+    # save using WriteVTK
+    path = joinpath(TEST_EXAMPLES_DIR, "spiral")
+    vtk_grid(path, points, polys) do vtk
+      vtk["theta", VTKPointData()] = point_values # scalar field attached to points
+      vtk["h", VTKCellData()] = poly_values    # scalar field attached to cells
```
```suggestion
      vtk["theta", VTKPointData()] = point_values # scalar field attached to points
      vtk["h", VTKCellData()] = poly_values       # scalar field attached to cells
```

ranocha commented on test/runtests.jl:
```diff
@@ -257,6 +259,54 @@ mkpath(TEST_EXAMPLES_DIR)
+    # check getter functions
+    @testset "get points" begin
+      _points = get_points(vtk)
+      @test all(points .== _points)
```
Do you need
```julia
@test all(points .== _points)
```
or does
```julia
@test points == _points
```
also work? If so, I would prefer the latter version (also two more times below).

ranocha commented on src/ReadVTK.jl:
```diff
@@ -141,6 +141,10 @@ function VTKFile(filename)
+  elseif file_type == "PolyData"
+    piece = root[file_type][1]["Piece"][1]
+    n_points = parse(Int, attribute(piece, "NumberOfPoints", required=true))
+    n_cells = 0
```
In the tests, you also load cell data. So there must be some way to get the number of cells, isn't there? But as far as I understand, cell data (such as the polygon data you used in the tests) is not required and there could be several kinds of additional data besides the points. Since it works in your case and you would like to use it soon, we could proceed as follows: 
What about adding some comments here about this issue and setting `c_cells= typemin(Int)`? This is obviously a value that must be wrong and doesn't necessarily indicate that there are no cells (such as `n_cells = 0` right now).

OndrejKincl replied:
I set n_cells = number of vertices + number of lines + number of polygons (which I think makes sense). I have not found a way how to obtain this number directly.

ranocha replied:
I would have associated the number of cells with higher-dimensional constructs, i.e., not vertices and points when there are polygons. However, I am not a VTK expert, so I don't know their manual by heart. What about this:
- You revert the change to `n_cells` (and `get_integer`, which is only required here, I think?) and set 
  ```julia
  # TODO: decide how to handle the number of cells correctly, see  
  #       https://github.com/trixi-framework/ReadVTK.jl/pull/11
  n_cells = typemin(Int)
  ```
  there.
- Then, I will have a final look, merge this PR soon, and release a new version that you can use for your application motivating you to make this PR.
- You create a new (draft) PR with the changes to `n_cells` here and we will wait until our VTK expert @sloede is available again.

ranocha replied:
Okay, let's keep it this way.

ranocha reviewed (changes requested):
Thanks a lot for the PR! Please find some comments below.

- [x] Could you please also update the README.md and docs/src/index.md files with the new file type information (in the sections what works, what does not work)? To be on the safe side, you could add that it works with some `PolyData`.
- [x] Could you please add your name to the list of contributors in the same files?

OndrejKincl commented:
I made several more changes:
* I set n_cells for PolyData to be sum of vertices + lines + polygons
* I realized that cell information cannot be obtained using the `get_cells` function. This is because PolyData can have cells of different kinds and have no attribute "Types". So I made separate struct for them, which I called `VTKPrimitives` and a function `get_primitives` to extract them. In case you don't like it, I can return these changes.
* I added another test for PolyData
* I updated the README

ranocha commented on README.md:
```diff
@@ -74,16 +74,18 @@ Further example VTK files can be found in the
 * Extracting point coordinates
 * Extracting information about cell types
 * Only for `ImageData` files: get origin, spacing, and extent information
+* Reading `PolyData` files containing vortices, lines and/or polygons
```
Could you please copy these changes also to `docs/src/index.md`?

OndrejKincl commented:
Thank you very much for being a reviewer. I believe it is not easy.
